Add DINOv2 with registers (using diff)

NielsRogge opened the pull request:
# What does this PR do?

This PR is a follow-up of #31832 but uses the diff instead of rewriting from scratch.

Had 2 issues with the diff:

- [ ] the `drop_path` function was causing an issue. Applying the suggestion at [#31832](https://github.com/huggingface/transformers/pull/31832#issuecomment-2298687531) fixed it.
- [ ] the diff doesn't properly add `super().__init__()` to the Backbone class. I manually added it in the modeling file as otherwise there's an error.

NielsRogge committed:
First draft

NielsRogge committed:
Convert model

NielsRogge committed:
Add docs

NielsRogge committed:
Fix index

NielsRogge committed:
Make fixup

NielsRogge committed:
Fix tests

NielsRogge committed:
First draft

NielsRogge committed:
Fix diff

NielsRogge committed:
Fix merge

NielsRogge committed:
Fix merge

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_32905). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

ArthurZucker commented on src/transformers/models/dinov2_with_registers/modeling_dinov2_with_registers.py:
was this actually generate with the diff converted? (I won't review it in this case!)

NielsRogge replied:
Yes it was

ArthurZucker commented on src/transformers/models/dinov2_with_registers/diff_dinov2_with_registers.py:
```diff
@@ -0,0 +1,376 @@
+        patch_pos_embed = patch_pos_embed.reshape(1, int(math.sqrt(num_positions)), int(math.sqrt(num_positions)), dim)
+        patch_pos_embed = patch_pos_embed.permute(0, 3, 1, 2)
+        target_dtype = patch_pos_embed.dtype
+        patch_pos_embed = nn.functional.interpolate(
+            patch_pos_embed.to(dtype=torch.float32),
+            scale_factor=(float(height / math.sqrt(num_positions)), float(width / math.sqrt(num_positions))),
+            mode="bicubic",
+            align_corners=False,
+            antialias=self.config.interpolate_antialias,
+        ).to(dtype=target_dtype)
```
can be split!

ArthurZucker commented on src/transformers/models/dinov2_with_registers/diff_dinov2_with_registers.py:
```diff
@@ -0,0 +1,376 @@
+        embeddings = embeddings + self.interpolate_pos_encoding(embeddings, height, width)
+
+        if self.register_tokens is not None:
+            embeddings = torch.cat(
+                (
+                    embeddings[:, :1],
+                    self.register_tokens.expand(embeddings.shape[0], -1, -1),
+                    embeddings[:, 1:],
+                ),
+                dim=1,
+            )
```
same here to split

ArthurZucker commented on src/transformers/models/dinov2_with_registers/diff_dinov2_with_registers.py:
```diff
@@ -0,0 +1,376 @@
+        # add positional encoding to each token
+        embeddings = embeddings + self.interpolate_pos_encoding(embeddings, height, width)
+
+        if self.register_tokens is not None:
```
tokens should always be register as this is register dino2!

ArthurZucker reviewed (commented):
Super nice! thanks for using diff, really fits well

NielsRogge committed:
Address comments

NielsRogge committed:
Address comment

NielsRogge commented:
@ArthurZucker thanks for the review, addressed your comments. Note that there were still some issues with the diff converter (see my first message), but other than that this PR seems ready

codybum commented:
Support for dinov2 with registers would be very much appreciated!

ArthurZucker commented:
#33248 will resolve all your issues 🤗

ArthurZucker commented:
I'll ask you to wait a tiny little bit as it will conflict with this work!

codybum commented:
> @ArthurZucker thanks for the review, addressed your comments. Note that there were still some issues with the diff converter (see my first message), but other than that this PR seems ready

Looks like Modular transformers: modularity and inheritance for new model additions https://github.com/huggingface/transformers/pull/33248 was merged a few hours ago!

ArthurZucker commented:
Yep! Make sure to rebase!

codybum commented:
I think I did the wrong @ initially. @NielsRogge do you have time to rebase and update?  If not, we can give it a go. 

> > @ArthurZucker thanks for the review, addressed your comments. Note that there were still some issues with the diff converter (see my first message), but other than that this PR seems ready
> 
> Looks like Modular transformers: modularity and inheritance for new model additions #33248 was merged a few hours ago!

NielsRogge commented:
Hi, feel free to take over, it's probably easy as only one method needs to be overwritten.

NielsRogge commented:
Closing this one, will try out the new modular Transformers when I have time
